### DISCUSSION TITLE
How to modify outputPath via Template / Transform

### DISCUSSION THREAD

**Discussion by whoisvadym** [11ty/eleventy#2494] [Q&A]
Hi, I'm an author of the [`eleventy-plugin-postcss`](https://github.com/pewstiepoll/eleventy-plugin-postcss) and as a part of my plugin functionality, I'd like to be able to modify the output path of the files that are created by my custom template.

The plugin works by attaching the extension via `.eleventy.js`:
```js
eleventyConfig.addExtension("<format>", { async compile() {} }
```
So far I've investigated the following options:
 1. I thought I could modify the `compile` method to return a modified outputPath pass along with string content, but it looks like the method is expected to return the string or Buffer.
 2. I was also thinking about modifying the output path by applying a Transform via `eleventyConfig.addTransform("<transform_name>", (_content, outputPath) => { ... });` but it looks like the `outputPath` is read-only within the transform function.

## Why
One of the primary reasons for this is to implement a "hash" functionality that would attach a random configurable hash to the output file name to enable a cache-busting functionality: `output-dir/../output-file-name.<hash>.css`

**Comment by zachleat** [comment]
https://www.11ty.dev/docs/languages/custom/#compileoptions.permalink-to-override-permalink-compilation is what you want I think!

**Reply by whoisvadym** [reply]
This looks promising, thanks @zachleat, I'll give it a try and get back to you!

**Reply by whoisvadym** [reply]
@zachleat Is there a way to get a default permalink and then apply my modifications on top of it in a custom `compileOptions.permalink` function? It looks like I'll need to implement a bunch of generic stuff like composing the outputPath along with applying custom modifications that I'm interested in.

Would be nice if the `permalink` callback could have the following third param: `(contents, inputPath, outputPath)  => {}` that way a custom permalink callback wouldn't have to compose the output path and could just reuse the result created by a default algorithm.

Would you consider a PR for this? I'd be glad to contribute if you find this useful.
